### PR TITLE
Add JSDoc based types

### PR DESCRIPTION
**Pull request by wooorm**
PR depends on work on `components` branch (https://github.com/remarkjs/react-markdown/pull/563).

### CODE REVIEW

File: src/rehype-filter.js [hunk 1]
```diff
@@ -4,6 +4,26 @@ const splice = [].splice
...
+ * @typedef {import('hast').Element} Element
+ *
+ * @callback AllowElement
+ * @param {Element} element
+ * @param {number} index
+ * @param {Element|Root} parent
+ * @returns {boolean}
+ *
+ * @typedef {Object} RehypeFilterOptions
+ * @property {Array.<string>} [allowedElements]
```
Comment (ChristianMurphy) [inline src/rehype-filter.js:18]: I personally find `string[]` to be a bit easier to read.
But am fine either way :+1:

**Review by ChristianMurphy — APPROVED** [review]
this one sparks joy

File: src/ast-to-react.js [hunk 2]
```diff
@@ -2,50 +2,205 @@
...
 
-    if (element(child)) {
+    if (child.type === 'element') {
       children.push(toReact(context, child, childIndex, node))
-    } else if (text(child)) {
+    } else if (child.type === 'text') {
       children.push(child.value)
-    } else if (child.type === 'raw' && !context.options.skipHtml) {
+    }
+    // @ts-ignore `raw` nodes are non-standard
```
[the remark this answers is not in the document]
Comment (wooorm) [reply src/ast-to-react.js:176]: true but they’re still “non-standard” (as in, not in `@types/hast`)

File: package.json [hunk 3]
```diff
@@ -175,6 +176,16 @@
     "trailingComma": "none",
     "useTabs": false
   },
+  "typeCoverage": {
+    "atLeast": 100,
+    "detail": true,
+    "strict": true,
+    "ignoreCatch": true,
```
[the remark this answers is not in the document]
Comment (wooorm) [reply package.json:183]: See second paragraph on something needed in TS in this section: https://github.com/plantain-00/type-coverage#ignore-catch.

File: src/react-markdown.js [hunk 4]
```diff
@@ -31,13 +56,16 @@ function ReactMarkdown(options) {
     .use(options.rehypePlugins || [])
     .use(filter, options)
 
+  /** @type {Root} */
+  // @ts-ignore we’ll throw if it isn’t a root next.
```
[the remark this answers is not in the document]
Comment (wooorm) [reply src/react-markdown.js:60]: I don’t think there’s a JSDoc equivalent, but you’re the expert here so let me know if I can solve it another way 😉

File: src/rehype-filter.js (hunk 1 above)
Comment (wooorm) [reply src/rehype-filter.js:18]: I tried both (in other projects). The shorter syntax is a bit nice. But I’m defaulting to JSDoc/Closure syntax (https://jsdoc.app/tags-type.html#overview) now in the hope of in the future using JSDoc or other tools to handle those annotations, where those tools might crash on TS syntax.

**Pull request merged**